modules/gateway/peers.go

```diff
@@ -109,27 +120,53 @@ func (g *Gateway) randomInboundPeer() (modules.NetAddress, error) {
 	return "", errNoPeers
 }
 
-// listen handles incoming connection requests. If the connection is accepted,
+// threadedListen handles incoming connection requests. If the connection is accepted,
 // the peer will be added to the Gateway's peer list.
-func (g *Gateway) listen() {
+func (g *Gateway) threadedListen() {
+	if g.threads.Add() != nil {
+		return
+	}
+	defer g.threads.Done()
+
+	// Shutdown the listener when the stop signal is received.
+	if g.threads.Add() != nil {
+		return
+	}
+	go func() {
+		defer g.threads.Done()
+		<-g.threads.StopChan()
+		if err := g.listener.Close(); err != nil {
+			g.log.Printf("WARN: listener.Close failed: %v", err)
+		}
+	}()
```

Comment:
I prefer having `g.threads.Stop()` be the first thing that you do when shutting down. It doesn't need to be the sole way of closing things, but in this case it's the only way to guarantee that `listener.Close` doesn't get called multiple times.